optimum/utils/normalized_config.py

```diff
@@ -102,6 +102,30 @@ class NormalizedVisionConfig(NormalizedConfig):
     INPUT_SIZE = "input_size"
 
 
+class NormalizedSegformerConfig(NormalizedVisionConfig):
+    NUM_ATTENTION_HEADS = "num_attention_heads"
+    HIDDEN_SIZE = "decoder_hidden_size"
+
+    def __getattr__(self, attr_name):
+        if attr_name == "hidden_size":
+            attr_value = getattr(self.config, self.HIDDEN_SIZE, None)
+            if attr_value is None:
+                raise AttributeError(f"Attribute {self.HIDDEN_SIZE} not found in config")
+            if isinstance(attr_value, list):
+                return max(attr_value)
+            return attr_value
+
+        elif attr_name == "num_attention_heads":
+            attr_value = getattr(self.config, self.NUM_ATTENTION_HEADS, None)
+            if attr_value is None:
+                raise AttributeError(f"Attribute {self.NUM_ATTENTION_HEADS} not found in config")
+            if isinstance(attr_value, list):
+                return max(attr_value)
```

Comment:
I went with max here, but if needed I can try to figure out how to implement list support

Reply:
I think it makes more sense to return the list as is. and it's the exporter's job to handle it. 
@mht-sharma do you know what `hidden_size` or `num_attention_heads` can be used for with image classification models ?

Reply:
So if I remove the `max` and run `pytest tests/onnxruntime/test_optimization.py -k test_compare_original_image_model_with_optimized_model -vv`, I get the following error:

```python
../../venvs/optimum/lib/python3.11/site-packages/onnxruntime/transformers/optimizer.py:178: in optimize_by_fusion
    optimizer = optimizer_class(model, num_heads, hidden_size)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <onnx_model_bert.BertOnnxModel object at 0x16fbebf50>, model = ir_version: 7
opset_import {
  version: 12
}
opset_import {
  domain: "com.microsoft.experimental"
  version: 1
}
opse...m_param: "batch_size"
          }
          dim {
            dim_value: 150
          }
        }
      }
    }
  }
}
, num_heads = [1, 2, 5, 8], hidden_size = 256

    def __init__(self, model: ModelProto, num_heads: int = 0, hidden_size: int = 0):
        """Initialize BERT ONNX Model.
    
        Args:
            model (ModelProto): the ONNX model
            num_heads (int, optional): number of attention heads. Defaults to 0 (detect the parameter automatically).
            hidden_size (int, optional): hidden dimension. Defaults to 0 (detect the parameter automatically).
        """
>       assert (num_heads == 0 and hidden_size == 0) or (num_heads > 0 and hidden_size % num_heads == 0)
E       TypeError: '>' not supported between instances of 'list' and 'int'
```

Reply:
It looks like the `onnxruntime` optimizer expects `num_attention_heads` to be an int.

Reply:
maybe https://github.com/microsoft/onnxruntime/issues/17254 is related

Reply:
https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/tools/transformers/onnx_model_bert.py#L37

The `BertOnnxModel` class can't handle a list for the number of attention heads.  Is BertOnnxModel the right class to use for the segformer model?

Reply:
HI @zachmayer it seems you are trying to optimise the `Segformer` for `image-classification` task. The model class for [SegformerModelForImageClassification](https://github.com/huggingface/transformers/blob/v4.40.1/src/transformers/models/segformer/modeling_segformer.py#L573) does not use the `decoder_hidden_size`. Not sure how the optimiser worked with `decoder_hidden_size`. Could you check the before and after graph if you see any changes?

Additionally, I am not super familiar with the architecture but on a quick read, the model uses hierarchal Transformer encoders, which explains multiple `hidden_sizes` and `attention_heads`. So this model may not be supported by onnxruntime itself.

I would suggest passing correct values, of hidden_size and heads, for a particular encoder block to the optimiser and see changes in the corresponding part in the graph.

Reply:
@mht-sharma — I'd actually like to optimise the `Segformer` for the semantic segmentation task! I was just following the patter for the test for the `vit` model.  I will update the test to use `ORTModelForSemanticSegmentation`.

The onnx runtime does support this model: I am able to:
- export it to onnx
- quantize it
- reload it and run inference

Here is an example of loading this model in onnx and quantizing it.  Everything seems to work great!
```python
from io import BytesIO
from pathlib import Path

import requests
from PIL import Image

from optimum.onnxruntime import ORTModelForSemanticSegmentation, ORTQuantizer
from optimum.onnxruntime.configuration import AutoQuantizationConfig


# Load a segformer
model_name = "hf-internal-testing/tiny-random-segformer"
model = ORTModelForSemanticSegmentation.from_pretrained(model_name, export=True)

# Quanzize
quantized_model_path = Path("~/Downloads/onnx_models/").expanduser() / model_name
quantizer = ORTQuantizer.from_pretrained(model)
quantizer.quantize(
    save_dir=quantized_model_path,
    quantization_config=AutoQuantizationConfig.avx2(is_static=False, per_channel=True),
)
model_quantized = ORTModelForSemanticSegmentation.from_pretrained(quantized_model_path)
del model

# Use the quantized model for inference on a test image
url = "https://upload.wikimedia.org/wikipedia/en/7/7d/Lenna_%28test_image%29.png"
response = requests.get(url)
image = Image.open(BytesIO(response.content))
processor = model_quantized.preprocessors[0]
inputs = processor.preprocess(images=image, return_tensors="np")
inputs = {"pixel_values": inputs["pixel_values"]}

# Run inference
outputs = model_quantized.model.run(None, inputs)[0]
print(outputs)
assert outputs.shape[1] == len(model_quantized.config.id2label)
```

Reply:
Given that the onnx runtime already supports `ORTModelForSemanticSegmentation`, and given that `ORTModelForSemanticSegmentation` supports quantizing, I thought we should also be able to support optimizing for this model.

Reply:
When I look at the config for the quantized model, I see it has a decoder_hidden_size:
```python
In [17]: print(model_quantized.config.decoder_hidden_size)
256
```

It also has a list of `hidden_sizes` and a list of `num_attention_heads`:
```python
In [18]: print(model_quantized.config.hidden_sizes)
[1, 2, 5, 8]

```

```python
In [19]: print(model_quantized.config.num_attention_heads)
[1, 2, 5, 8]
```

Reply:
So Microsoft hasn't tested the optimizer with segformers yet: https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/python/tools/transformers/README.md#supported-models

They expect both num_heads and hidden_size to be ints:
```
num_heads (int, optional): number of attention heads. Defaults to 0.
        0 allows detect the parameter from graph automatically.
    hidden_size (int, optional): hidden size. Defaults to 0.
        0 allows detect the parameter from graph automatically.
```
        
If I convert `model_quantized.config.hidden_sizes` and `model_quantized.config.num_attention_heads` to integers by taking the max, the optimizer works and does seem to change the graph of the model

Reply:
@zachmayer,  curious to know which part of the model / encoder block is optimised. Since you gave max values I suppose the last one in the hierarchy? Or all encoder blocks are optimised?

If first, then it could be worth to try out having a loop on the optimiser for each pair of hidden size / attention head and see if all the blocks are optimised?

Reply:
How do I tell which blocks get optimized?

I just pushed a commit that uses sum instead of max, and uses the `ORTModelForSemanticSegmentation` in the test, and the test passes.

I'll try some different parameters and see what changes.

Reply:
num_heads and hidden_size is optional for onnxruntime transformers optimizer. If you are not sure, set them to 0.

You can use Netron to view the onnx model before/after optimization.

Reply:
@tianleiwu — ok!  I just pushed a new commit where I set them to 0.